### PR TITLE
chore: add ci analysis for sonar

### PR DESCRIPTION
**Pull request by bmunguli**
## Description

- Run CI analysis with sonar instead of automatic analysis which gives the opportunity for publishing the test coverage results in sonar cloud
- Use maven plugin to integrate with sonar  as this is  the recommended way
- Add jacoco to produce test coverage reports and show results in sonarcloud 

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [x] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [x] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -98,9 +98,20 @@
...
+		<jacoco.version>0.8.11</jacoco.version>
+
+		<!-- sonar plugin -->
+		<sonar-maven-plugin.version>3.10.0.2594</sonar-maven-plugin.version>
+		
 
 		<!-- build -->
 		<maven.compiler.version>3.8.1</maven.compiler.version>
+
+
```
Comment (tuncaytunc-zf) [inline pom.xml:110]: Please remove the second empty line, one empty line should be enough.

**Review by tuncaytunc-zf — APPROVED** [review]
LGTM

**Comment by tunacicek** [comment]
Hi @bmunguli @tuncaytunc-zf ,
the build failed. Do you have the possibility to see the logs?

**Comment by tunacicek** [comment]
> > Can someone with sonar cloud access rights disable the automatic analysis on sonar so the cli integration can be fully tested if this PR looks good and we would like to merge it? Currently the pipeline is failing because of conflicts between automatic analysis and ci analysis.
> > Refs: https://docs.sonarsource.com/sonarcloud/advanced-setup/automatic-analysis/#conflict-with-ci-based-analysis
> 
> @tunacicek thanks for reviewing this, as mentioned in my previous comment the pipeline is currently failing because of conflicts between automatic analysis and ci analysis analysis enabled in sonarcloud. I need someone with access rights in tractus-x sonarcloud project https://sonarcloud.io/project/overview?id=eclipse-tractusx_sldt-digital-twin-registry to disable the configured automatic analysis so the ci analysis could work.
> 
> From the logs:
> 
> `Error: Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.10.0.2594:sonar (default-cli) on project digital-twin-registry: You are running CI analysis while Automatic Analysis is enabled. Please consider disabling one or the other. -> [Help 1]`
> 
> Refs: https://docs.sonarsource.com/sonarcloud/advanced-setup/automatic-analysis/#conflict-with-ci-based-analysis

@bmunguli : Thanks for your feedback.  
@Siegfriedk, @RoKrish14 : Could one of you please disable the configured automatic analysis?

**Comment by tunacicek** [comment]
> @tunacicek running the dash tool via maven plugin locally updates the DEPENDENCIES file with some libraries i have not introduced as part of this PR. The two new plugins that are introduced: `jacoco` and `sonar` are not reflected because they are not dependencies. Screenshot 2024-03-05 at 11 13 22
> 
> Should I skip including those changes as part of this PR?

@bmunguli : yes, you can skip those changes. thank you

**Comment by Siegfriedk** [comment]
@tunacicek i disabled the automated analysis

**Comment by tunacicek** [comment]
> @tunacicek i disabled the automated analysis

@Siegfriedk :Thank you :)

**Comment by tunacicek** [comment]
> @tunacicek can you please check if the SONAR_TOKEN is set as a secret on this repository or organisation level?

@bmunguli  i have no permission to check it. 
@Siegfriedk : Could you please check if the token is set? 

@bmunguli : A reference implementation can be found here: 
https://github.com/eclipse-tractusx/traceability-foss

**Comment by Siegfriedk** [comment]
@tunacicek only EF can check it. For adhoc, you can ask frederric gurr per matrix and hope he doesn't bite you :)

**Comment by tunacicek** [comment]
@Siegfriedk thanks, I will try it :)

**Comment by fredg02** [comment]
Neither a repo-level (for `sldt-digital-twin-registry`) nor an org-level `SONAR_TOKEN` secret exists.

**Comment by fredg02** [comment]
> @tunacicek only EF can check it. For adhoc, you can ask frederric gurr per matrix and hope he doesn't bite you :)

1:1 communication does not scale. So please rather mention me in a comment here or create a HelpDesk issue.

**Comment by tunacicek** [comment]
> Neither a repo-level (for `sldt-digital-twin-registry`) nor an org-level `SONAR_TOKEN` secret exists.

@fredg02 : thanks for your feedback. I would create a ticket to have a token.

**Comment by tunacicek** [comment]
@bmunguli : I created the ticket and it is approved/ created. 
https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/3299#note_1156633

I rerun the build but it failed. Could you add the correct projectid?

**Comment by tunacicek** [comment]
> @bmunguli : I created the ticket https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/4399

@bmunguli ticket is closed. SONAR_TOKEN is available now. I rerun the build

File: .github/workflows/build-snapshot.yml [hunk 2]
```diff
@@ -39,13 +39,28 @@ jobs:
...
         env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
\ No newline at end of file
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: SonarCloud Code Analysis
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}  # Needed to get PR information, if any
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+        run: mvn -B --batch-mode sonar:sonar -Dsonar.projectKey=eclipse-tractusx_sldt-digital-twin-registry
```
Comment (tuncaytunc-zf) [inline .github/workflows/build-snapshot.yml:66]: I think the ``` -Dsonar.organization=eclipse-tractusx ``` parameter is missing here.

**Comment by tunacicek** [comment]
> @tunacicek from the logs its complaining that: `failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.10.0.2594:sonar (default-cli) on project digital-twin-registry: Project not found. Please check the 'sonar.projectKey' and 'sonar.organization' properties, the 'SONAR_TOKEN' environment variable, or contact the project administrator `
> 
> Double Checking the projectKey and organization look correct to me and matching with the project in sonar. The SONAR_TOKEN is usually bound to a user, could it be that the user who generated the token does not have permissions to perform analysis on this project?

Hi @pstankie, Could you check it please?

**Comment by pstankie** [comment]
I added SONAR_ORGANIZATION and SONAR_PROJECT_KEY github variables to repository settings. You can use them:

    orgs.newRepoVariable('SONAR_ORGANIZATION') {
          value: "eclipse-tractusx",
        },
        orgs.newRepoVariable('SONAR_PROJECT_KEY') {
          value: "eclipse-tractusx_sldt-digital-twin-registry",
        },

**Comment by tunacicek** [comment]
> I added SONAR_ORGANIZATION and SONAR_PROJECT_KEY github variables to repository settings. You can use them:
> 
> ```
> orgs.newRepoVariable('SONAR_ORGANIZATION') {
>       value: "eclipse-tractusx",
>     },
>     orgs.newRepoVariable('SONAR_PROJECT_KEY') {
>       value: "eclipse-tractusx_sldt-digital-twin-registry",
>     },
> ```

@pstankie :Thank you very much for your support. 
@bmunguli : Could you please use the variables instead of hard coded values.

**Comment by pstankie** [comment]
I updated SONAR_TOKEN. Can you try again ?

**Comment by pstankie** [comment]
So the SONAR_TOKEN is not visible in  PR from forked repository. You may try `pull_request_target` trigger in this specific case.

**Comment by tunacicek** [comment]
> @tunacicek as @pstankie commented previously github does not allow access to secrets on forked PRs so the sonar analysis can not happen there. I asked the question on the [community](https://chat.eclipse.org/#/room/!DWDObjgSCMTOqevBVY:matrix.eclipse.org/$hqfcSaVVVgauI-rl_tyTbYVmCdqKbhttdzOOFGlUbLw?via=matrix.eclipse.org&via=matrix.org&via=matrix.diverse-team.fr). Seems there are few options recommended:
> 
> * changing to pull_request_target
> * using a github app for the integration
> * switching on again automatic sonar analysis as the support from sonar is not that great
> * https://github.com/eclipse-tractusx/traceability-foss/actions/runs/8319310761/workflow or running the scans only in the main branch as they do in this reference implementation
> * Also for commiters this will work if they create PRs directly into the main branch
> 
> I am not super confident about the setup of the first two, shall we go ahead and use same approach as [traceability-foss](https://github.com/eclipse-tractusx/traceability-foss) for now (the drawback with this is that we can not get feedback if the integration works until changes got to main)?

Hi @bmunguli , 
yes, we can do it like in traceability and then check whether it works or not. Could you fix the merge conflicts?

**Review by tunacicek — APPROVED** [review]
LGTM

**Pull request merged**